v4/newrelic/transaction.go

```diff
@@ -214,11 +230,40 @@ func (txn *Transaction) InsertDistributedTraceHeaders(hdrs http.Header) {}
 // AcceptDistributedTraceHeaders first looks for the presence of W3C trace
 // context headers.  Only when those are not found will it look for the New
 // Relic distributed tracing header.
-func (txn *Transaction) AcceptDistributedTraceHeaders(t TransportType, hdrs http.Header) {}
+func (txn *Transaction) AcceptDistributedTraceHeaders(t TransportType, hdrs http.Header) {
+	if txn == nil {
+		return
+	}
+
+	txn.thread.currentSpan.Lock()
+	defer txn.thread.currentSpan.Unlock()
+
+	// Here we create a OpenTelemetry context that is detached from the
+	// current trace. All segments (spans) subsequently started with this
+	// context will be detached from the transaction trace, but rather will
+	// have the remote trace id as trace id and the remote span id as the
+	// parent span id.
+	//
+	// If no more than the root segment where yet created for this
+	// transaction, we discard the root segment and replace it with a new
+	// root segment that has the proper remote parent and trace id.
+	remoteCtx := propagation.ExtractHTTP(context.Background(), txn.app.propagators, hdrs)
+
+	if txn.thread.spanCount == 1 {
+		ctx, sp := txn.app.tracer.Start(remoteCtx, txn.name)
+		txn.rootSpan = &span{
+			Span: sp,
+			ctx:  ctx,
+		}
+		txn.thread.currentSpan = txn.rootSpan
+	} else {
+		txn.thread.currentSpan.ctx = remoteCtx
```

Comment:
I think all the swapping of currentSpan and rootSpan here should be done under lock.

Reply:
It is, see [line 238](https://github.com/newrelic/go-agent/pull/191/files#diff-6b77bc7503ec9881e453708a4bb0f3b1R238). I used the mutex on `txn.thread.currentSpan`. I'm not sure if this merits introducing another mutex on the transaction itself.

Reply:
Ohhh duh! I'm blind. Thanks!